substrate/frame/identity/src/lib.rs

```diff
@@ -925,7 +1179,99 @@ impl<T: Config> Pallet<T> {
+	/// the characters in `username` and that `length` (if `Some`) conforms to the limit. It is not
+	/// expected to pass a fully formatted username here (i.e. one with any protocol-added
+	/// characters included, such as a `.`).
+	fn validate_username(username: &Vec<u8>, length: Option<u32>) -> DispatchResult {
+		// Verify input length before allocating a Vec with the user's input. `<` instead of `<=`
+		// because it needs one element for the point (`username` + `.` + `suffix`).
+		if let Some(l) = length {
+			ensure!(l < USERNAME_MAX_LENGTH, Error::<T>::InvalidUsername);
+		}
+		ensure!(
+			username.iter().all(|byte| byte.is_ascii_alphanumeric()),
+			Error::<T>::InvalidUsername
+		);
+		Ok(())
+	}
+
+	/// Validate a signature. Supports signatures on raw `data` or `data` wrapped in HTML `<Bytes>`.
+	pub fn validate_signature(
+		data: &Vec<u8>,
+		signature: &T::OffchainSignature,
+		signer: &T::AccountId,
+	) -> DispatchResult {
+		// Happy path, user has signed the raw data.
+		if signature.verify(&data[..], &signer) {
+			return Ok(())
+		}
+		// NOTE: for security reasons modern UIs implicitly wrap the data requested to sign into
+		// `<Bytes> + data + </Bytes>`, so why we support both wrapped and raw versions.
+		let prefix = b"<Bytes>";
+		let suffix = b"</Bytes>";
+		let mut wrapped: Vec<u8> = Vec::with_capacity(data.len() + prefix.len() + suffix.len());
+		wrapped.extend(prefix);
+		wrapped.extend(data);
+		wrapped.extend(suffix);
+
+		ensure!(signature.verify(&wrapped[..], &signer), Error::<T>::InvalidSignature);
+
+		Ok(())
+	}
+
+	/// A username has met all conditions. Insert the relevant storage items.
+	pub fn insert_username(who: &T::AccountId, username: Username) {
+		// Check if they already have a primary. If so, leave it. If not, set it.
+		// Likewise, check if they have an identity. If not, give them a minimal one.
+		let (reg, primary_username, new_is_primary) = match <IdentityOf<T>>::get(&who) {
+			// User has an existing Identity and a primary username. Leave it.
+			Some((reg, Some(primary))) => (reg, primary, false),
+			// User has an Identity but no primary. Set the new one as primary.
+			Some((reg, None)) => (reg, username.clone(), true),
+			// User does not have an existing Identity. Give them a fresh default one and set
+			// their username as primary.
+			None => (
+				Registration {
+					info: T::IdentityInformation::default(),
+					judgements: Default::default(),
+					deposit: Zero::zero(),
+				},
+				username.clone(),
+				true,
+			),
+		};
+
+		// Enter in identity map. Note: In the case that the user did not have a pre-existing
+		// Identity, we have given them the storage item for free. If they ever call
+		// `set_identity` with identity info, then they will need to place the normal identity
+		// deposit.
+		IdentityOf::<T>::insert(&who, (reg, Some(primary_username)));
+		// Enter in username map.
+		AccountOfUsername::<T>::insert(username.clone(), &who);
```

Comment:
Shouldnt this only be done if `new_is_primary`?

Reply:
No, primary refers to the username that `IdentityOf` maps to, i.e. the `AccountId => Username` mapping. But multiple usernames can map to the same `AccountId`.